File: backend/src/main/kotlin/ru/sidnystorm/books/service/UserService.kt
package ru.sidnystorm.books.service

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import ru.sidnystorm.books.dto.User
import ru.sidnystorm.books.dto.UserInfo
import ru.sidnystorm.books.dto.UserRole
import ru.sidnystorm.books.exception.DuplicatedUserException
import ru.sidnystorm.books.repository.UserRepository

@Service
class UserService {

   @Autowired
   lateinit var repository: UserRepository
   @Autowired
   lateinit var encoder: PasswordEncoder

   fun add(user: User) : Int {
      if (repository.findAuth(user.login!!) != null) {
         throw DuplicatedUserException()
      }

      user.password = encoder.encode(user.password)
      user.role = UserRole.USER
      return repository.insert(user)
   }

   fun find(login: String) = repository.find(User(login))

   //TODO existence check
   fun update(login: String, userInfo: UserInfo) = repository.update(login, userInfo)
}

File: frontend/src/service/user.service.js
import API from './util';

const API_URL = '/user';

class UserService {

   addUser(userInfo) {
      return API
         .post(`${API_URL}/add`, {
            login: userInfo.login,
            password: userInfo.password,
            userInfo})
         .then(response => response);
   }

   findUser(login) {
      return API.post(`${API_URL}/find?login=${login}`)
         .then(response => response.data, error => error?.response);
   }

   update(login, userInfo) {
      return API
         .post(`${API_URL}/update?login=${login}`, { ...userInfo })
         .then(response => response.data, error => error.response);
   }
}

export default new UserService();